diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Bitwuzla.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Bitwuzla.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Bitwuzla.java	2024-01-30 02:17:29.958540211 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Bitwuzla.java	2024-01-30 02:13:50.700388020 +0100
@@ -27,7 +27,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Exception.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Exception.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Exception.java	2024-01-30 02:17:08.706597398 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Exception.java	2024-01-30 02:13:50.576389963 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Map_TermTerm.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Map_TermTerm.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Map_TermTerm.java	2024-01-30 02:16:53.494652530 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Map_TermTerm.java	2024-01-30 02:13:50.564390151 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
@@ -171,7 +171,7 @@
   
     @SuppressWarnings("deprecation")
     protected void finalize() {
-      delete();
+      //delete();
     }
   
     public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Options.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Options.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Options.java	2024-01-30 02:17:14.802578662 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Options.java	2024-01-30 02:13:50.596389650 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Parser.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Parser.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Parser.java	2024-01-30 02:17:40.450520255 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Parser.java	2024-01-30 02:13:50.704387957 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Sort.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Sort.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Sort.java	2024-01-30 02:17:20.062564011 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Sort.java	2024-01-30 02:13:50.636389023 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Terminator.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Terminator.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Terminator.java	2024-01-30 02:16:42.862698311 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Terminator.java	2024-01-30 02:13:50.640388960 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Term.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Term.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Term.java	2024-01-30 02:17:34.346531214 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Term.java	2024-01-30 02:13:50.624389211 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Int.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Int.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Int.java	2024-01-30 02:16:58.418633361 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Int.java	2024-01-30 02:13:50.460391781 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Sort.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Sort.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Sort.java	2024-01-30 02:17:24.998551530 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Sort.java	2024-01-30 02:13:50.532390653 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_String.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_String.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_String.java	2024-01-30 02:16:35.794732123 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_String.java	2024-01-30 02:13:50.484391405 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
diff -ru lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Term.java lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Term.java
--- lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Term.java	2024-01-30 02:17:03.534614791 +0100
+++ lib/native/source/libbitwuzla/src.patched/org/sosy_lab/java_smt/solvers/bitwuzla/api/Vector_Term.java	2024-01-30 02:13:50.508391029 +0100
@@ -35,7 +35,7 @@
 
   @SuppressWarnings("deprecation")
   protected void finalize() {
-    delete();
+    //delete();
   }
 
   public synchronized void delete() {
