Index: src/api/Interpret.cc
diff --git a/src/api/Interpret.cc b/src/api/Interpret.cc
--- a/src/api/Interpret.cc	(revision e4c8030849e6c2896735567b5d866fdccc70e9ae)
+++ b/src/api/Interpret.cc	(date 1774016994607)
@@ -27,6 +27,7 @@
 #include "Interpret.h"
 
 #include <api/smt2tokens.h>
+#include <common/ApiException.h>
 #include <logics/ArithLogic.h>
 #include <logics/LogicFactory.h>
 #include <rewriters/Substitutor.h>
@@ -42,7 +43,12 @@
 
 namespace opensmt {
 
-Interpret::~Interpret() = default;
+Interpret::~Interpret() {
+    // If the logic was given in the constructor, and not set by the string-input, we have to clean up the reference
+    if (!ownsLogic) {
+        logic.release();
+    }
+}
 
 PTRef
 Interpret::getParsedFormula()
@@ -1055,11 +1061,13 @@
 
 
 void Interpret::notify_formatted(bool error, const char* fmt_str, ...) const {
+    std::stringstream out;
+
     va_list ap;
     int d;
     char c1, *t;
     if (error) {
-        std::cout << "(error \"";
+        out << "(error \"";
         _okStatus = false;
         // also set `f_exit = true` if configured to stop on first error
     }
@@ -1071,25 +1079,29 @@
             switch (*fmt_str++) {
             case 's':
                 t = va_arg(ap, char *);
-                std::cout << t;
+                out << t;
                 break;
             case 'd':
                 d = va_arg(ap, int);
-                std::cout << d;
+                out << d;
                 break;
             case '%':
-                std::cout << '%';
+                out << '%';
                 break;
             }
         }
         else if (c1 != '\0')
-            std::cout << c1;
+            out << c1;
         else break;
     }
     va_end(ap);
     if (error)
-        std::cout << "\")";
-    std::cout << std::endl;
+        out << "\")";
+    out << std::endl;
+
+    if (error)
+        throw ApiException(out.str());
+    std::cout << out.str();
 }
 
 void Interpret::notify_success() const {
Index: src/api/Interpret.h
diff --git a/src/api/Interpret.h b/src/api/Interpret.h
--- a/src/api/Interpret.h	(revision e4c8030849e6c2896735567b5d866fdccc70e9ae)
+++ b/src/api/Interpret.h	(date 1774016742039)
@@ -138,6 +138,7 @@
     std::unique_ptr<MainSolver> main_solver;

     bool            f_exit;
+    bool            ownsLogic;
     mutable bool    _okStatus{true};

     vec<PTRef>      assertions;
@@ -189,7 +190,18 @@
     Interpret(SMTConfig & c)
         : config     (c)
         , f_exit     (false)
-        { }
+        , ownsLogic  (true)
+        { }
+
+    /** This constructor reuses the existing logic from caller context. */
+    Interpret(SMTConfig & c, Logic & l)
+        : config     (c)
+        , f_exit     (false)
+        , ownsLogic  (false) {
+        logic.reset(&l);
+        main_solver = createMainSolver(l.getName().c_str());
+        main_solver->initialize();
+    }

     ~Interpret();

Index: src/api/MainSolver.h
diff --git a/src/api/MainSolver.h b/src/api/MainSolver.h
--- a/src/api/MainSolver.h	(revision e4c8030849e6c2896735567b5d866fdccc70e9ae)
+++ b/src/api/MainSolver.h	(date 1774016994782)
@@ -112,7 +112,7 @@
     // I.e., *excluding* popped assertions
     vec<PTRef> getCurrentAssertions() const;
     // Returns just a view to the assertions
-    auto getCurrentAssertionsView() const { return getCurrentAssertionsViewImpl(); }
+    vec<PTRef> getCurrentAssertionsView() const { return getCurrentAssertionsViewImpl(); }
 
     vec<PTRef> const & getAssertionsAtCurrentLevel() const { return getAssertionsAtLevel(getAssertionLevel()); }
     vec<PTRef> const & getAssertionsAtLevel(std::size_t) const;
Index: src/common/numbers/FastRational.cc
diff --git a/src/common/numbers/FastRational.cc b/src/common/numbers/FastRational.cc
--- a/src/common/numbers/FastRational.cc	(revision e4c8030849e6c2896735567b5d866fdccc70e9ae)
+++ b/src/common/numbers/FastRational.cc	(date 1774016690334)
@@ -15,18 +15,22 @@
 mpq_ptr FastRational::mpqPool::alloc()
 {
     mpq_ptr r;
+    cs.lock();
     if (!pool.empty()) {
         r = pool.top();
         pool.pop();
     } else {
         r = store.emplace().get_mpq_t();
     }
+    cs.unlock();
     return r;
 }
 
 void FastRational::mpqPool::release(mpq_ptr ptr)
 {
+    cs.lock();
     pool.push(ptr);
+    cs.unlock();
 }
 
 FastRational::FastRational( const char * s, const int base )
Index: src/common/numbers/FastRational.h
diff --git a/src/common/numbers/FastRational.h b/src/common/numbers/FastRational.h
--- a/src/common/numbers/FastRational.h	(revision e4c8030849e6c2896735567b5d866fdccc70e9ae)
+++ b/src/common/numbers/FastRational.h	(date 1774016742090)
@@ -11,6 +11,7 @@
 #include <cassert>
 #include <climits>
 #include <cstdint>
+#include <mutex>
 #include <optional>
 #include <stack>
 #include <string>
@@ -75,6 +76,7 @@
 {
     class mpqPool
     {
+        std::mutex cs; // We use a mutex to make the pool thread-safe
         std::stack<mpq_class> store; // uses deque as storage to avoid realloc
         std::stack<mpq_ptr, std::vector<mpq_ptr>> pool;
     public:
Index: src/logics/Logic.cc
diff --git a/src/logics/Logic.cc b/src/logics/Logic.cc
--- a/src/logics/Logic.cc	(revision e4c8030849e6c2896735567b5d866fdccc70e9ae)
+++ b/src/logics/Logic.cc	(date 1774016994729)
@@ -9,6 +9,7 @@
 #include "Logic.h"
 #include "SubstLoopBreaker.h"
 
+#include <api/Interpret.h>
 #include <api/smt2tokens.h>
 #include <common/ApiException.h>
 #include <common/InternalException.h>
@@ -674,6 +675,18 @@
     return mkFun(sr, {});
 }
 
+PTRef Logic::parseFormula(std::string input) {
+    SMTConfig config = SMTConfig();
+    Interpret repl = Interpret(config, *this);
+
+    int rval = repl.interpFile(input.data());
+    if (rval != 0) {
+        throw ApiException("Failed to parse input formula");
+    }
+
+    return repl.getParsedFormula();
+}
+
 void Logic::instantiateFunctions(SRef sr) {
     // Equality
     SymRef tr = declareFun_Commutative_NoScoping_Chainable(tk_equals, sort_BOOL, {sr, sr});
@@ -746,10 +759,8 @@
         assert(terms.size() == 0);
         return mkFun(sym, std::move(terms));
     }
-    if (hasArrays()) {
-        if (isArrayStore(sym)) return mkStore(std::move(terms));
-        if (isArraySelect(sym)) return mkSelect(std::move(terms));
-    }
+    if (isArrayStore(sym)) return mkStore(std::move(terms));
+    if (isArraySelect(sym)) return mkSelect(std::move(terms));
     return mkUninterpFun(sym, std::move(terms));
 }
 
@@ -1347,6 +1358,9 @@
     }
 }
 
+Sort const&       Logic::getSortDefinition (SRef s)     const { return sort_store[s]; }
+SortSymbol const& Logic::getSortSymbol     (SSymRef ss) const { return sort_store[ss]; }
+
 // Fetching sorts
 SRef Logic::getSortRef(PTRef const tr) const {
     return getSortRef(getPterm(tr).symb());
Index: src/logics/Logic.h
diff --git a/src/logics/Logic.h b/src/logics/Logic.h
--- a/src/logics/Logic.h	(revision e4c8030849e6c2896735567b5d866fdccc70e9ae)
+++ b/src/logics/Logic.h	(date 1774016994762)
@@ -59,6 +59,9 @@
     bool hasIntegers() const { return QFLogicToProperties.at(logicType).arithProperty.hasInts; }
     bool hasReals() const { return QFLogicToProperties.at(logicType).arithProperty.hasReals; }
 
+    const Sort& getSortDefinition (SRef s) const;
+    const SortSymbol& getSortSymbol (SSymRef ss) const;
+
     SRef getSortRef(PTRef tr) const;
     SRef getSortRef(SymRef sr) const;
     std::string sortToString(SRef s) const;
@@ -222,6 +225,8 @@
     void dumpWithLets(std::ostream & out, PTRef formula) const;
     std::string dumpWithLets(PTRef formula) const;
 
+    PTRef parseFormula(std::string input);
+
     PTRef instantiateFunctionTemplate(TemplateFunction const & tmplt, vec<PTRef> const & args);
 
     SSymRef getSortSymIndexed() const { return sym_IndexedSort; }
Index: src/symbols/SymStore.cc
diff --git a/src/symbols/SymStore.cc b/src/symbols/SymStore.cc
--- a/src/symbols/SymStore.cc	(revision e4c8030849e6c2896735567b5d866fdccc70e9ae)
+++ b/src/symbols/SymStore.cc	(date 1774016742104)
@@ -26,6 +26,7 @@
 #include "SymStore.h"
 #include "Symbol.h"
 
+#include <common/ApiException.h>
 #include <cstring>
 
 namespace opensmt {
@@ -54,6 +55,10 @@
             }
         }
     }
+    if (symrefs != nullptr && !symConfig.isInterpreted) {
+        // Only builtin functions like "=" or "+" should be polymorphic
+        throw ApiException("Symbol `" + std::string(fname) + "` has conflicting definitions");
+    }
     bool newsym = (symrefs == nullptr);
     SymRef tr = ta.alloc(rsort, args, symConfig);
     SymId id = symbols.size();
