diff --git a/src/api/java/CMakeLists.txt
--- a/src/api/java/CMakeLists.txt
+++ b/src/api/java/CMakeLists.txt
@@ -159,8 +159,8 @@
 add_dependencies(cvc5jni generate-jni-headers)
 
-target_include_directories(cvc5jni PUBLIC ${JNI_INCLUDE_DIRS})
 target_include_directories(cvc5jni PUBLIC ${PROJECT_SOURCE_DIR}/src)
 target_include_directories(cvc5jni PUBLIC ${CMAKE_BINARY_DIR}/src/)
 target_include_directories(cvc5jni PUBLIC ${JNI_DIR})
-target_link_libraries(cvc5jni PRIVATE ${JNI_LIBRARIES})
+include_directories(${JNI_INCLUDE_DIRS})
+link_libraries(${JNI_LIBRARIES})
 target_link_libraries(cvc5jni PRIVATE cvc5)
