Index: src/main/java/com/sri/yices/Makefile
diff --git a/src/main/java/com/sri/yices/Makefile b/src/main/java/com/sri/yices/Makefile
--- a/src/main/java/com/sri/yices/Makefile	(revision 2be85e119b6004cc489172568398582268a4f59d)
+++ b/src/main/java/com/sri/yices/Makefile	(date 1773243769420)
@@ -56,9 +56,10 @@
 # we ignore versions and soname for now
 
 # default include directories for jni.h and jni_md.h
-CPPFLAGS := -I $(JAVA_HOME)/include -I $(JAVA_HOME)/include/$(OS)
+CPPFLAGS := -I $(JAVA_HOME)/include -I $(JAVA_HOME)/include/$(OS) -I $(GMP_PATH)/include -I $(YICES_PATH)/include
 CXXFLAGS := -g -fPIC
-LIBS := -lyices -lgmp
+LDFLAGS := -L $(YICES_PATH)/lib -L $(CUDD_PATH)/lib -L $(POLY_PATH)/lib -L $(GMP_PATH)/lib
+LIBS := -lyices -lcudd -lpoly -lgmpxx -lgmp
 
 CXX ?= g++
 
@@ -86,7 +87,7 @@
 	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -dynamiclib -o $@ yicesJNI.o $(LIBS)
 
 libyices2java.so: yicesJNI.o
-	$(CXX) $(CFLAGS) $(LDFLAGS) -shared -o $@ yicesJNI.o $(LIBS)
+	$(CXX) $(CFLAGS) $(LDFLAGS) -o $@ yicesJNI.o -shared -Wl,-soname,libyices2java.so -Wl,-Bstatic $(LIBS) -static-libstdc++ -lstdc++ -Wl,-Bdynamic -lc -lm -Wl,--version-script=libyices2java.version
 
 LIBDIR := $(YICES_JNI)
 
Index: src/main/java/com/sri/yices/libyices2java.version
diff --git a/src/main/java/com/sri/yices/libyices2java.version b/src/main/java/com/sri/yices/libyices2java.version
new file mode 100644
--- /dev/null	(date 1773151156723)
+++ b/src/main/java/com/sri/yices/libyices2java.version	(date 1773151156723)
@@ -0,0 +1,4 @@
+YICES2JAVA {
+  global: Java_*;
+  local: *;
+};
